drivers/serial/uart_sf32lb.c

```diff
@@ -351,6 +396,363 @@ static void uart_sf32lb_irq_rx_disable(const struct device *dev)
+	if (data->async.cb) {
+		data->async.cb(uart_dev, &evt, data->async.user_data);
+	}
+
+	/* Load next buffer and request another */
+	data->async.rx.buf = data->async.rx.next_buf;
+	data->async.rx.len = data->async.rx.next_len;
+	data->async.rx.next_buf = NULL;
+	data->async.rx.next_len = 0U;
+
+	/* If there's a next buffer, restart DMA */
+	if (data->async.rx.buf) {
+		sf32lb_dma_reload_dt(&config->rx_dma, (uintptr_t)(config->base + UART_RDR),
+				     (uint32_t)data->async.rx.buf, data->async.rx.len);
+
+		err = sf32lb_dma_start_dt(&config->rx_dma);
+		if (err) {
+			LOG_ERR("Error starting Rx DMA (%d)", err);
+		} else {
+			sys_set_bit(config->base + UART_CR3, USART_CR3_DMAR_Pos);
+			sys_set_bit(config->base + UART_RQR, USART_RQR_RXFRQ_Pos);
+		}
+	}
+
+	evt.type = UART_RX_BUF_REQUEST;
+	if (data->async.cb) {
+		data->async.cb(uart_dev, &evt, data->async.user_data);
+	}
+
+	/* Notify RX_DISABLED when there is no buffer */
+	if (!data->async.rx.buf) {
+		evt.type = UART_RX_DISABLED;
+		if (data->async.cb) {
+			data->async.cb(uart_dev, &evt, data->async.user_data);
+		}
+	}
+
+	irq_unlock(key);
+}
+
+static int uart_async_sf32lb_callback_set(const struct device *dev, uart_callback_t callback,
+					  void *user_data)
+{
+	struct uart_sf32lb_data *data = dev->data;
+
+	data->async.cb = callback;
+	data->async.user_data = user_data;
+
+	return 0;
+}
+
+static int uart_async_sf32lb_rx_enable(const struct device *dev, uint8_t *buf, size_t len,
+				       int32_t timeout)
+{
+	const struct uart_sf32lb_config *config = dev->config;
+	struct uart_sf32lb_data *data = dev->data;
+	int err = 0;
+	struct uart_event evt = {0};
+
+	if (config->rx_dma.channel == 0xFF) {
+		LOG_ERR("Rx DMA channel is not configured");
+		return -ENOTSUP;
+	}
+
+	unsigned int key = irq_lock();
+
+	data->async.rx.buf = buf;
+	data->async.rx.len = len;
+	data->async.rx.timeout = timeout;
+
+	sf32lb_dma_reload_dt(&config->rx_dma, (uintptr_t)(config->base + UART_RDR),
+			     (uint32_t)data->async.rx.buf, data->async.rx.len);
+
+	err = sf32lb_dma_start_dt(&config->rx_dma);
+	if (err) {
+		LOG_ERR("Error starting Rx DMA (%d)", err);
+		goto unlock;
+	}
+
+	sys_set_bit(config->base + UART_CR3, USART_CR3_DMAR_Pos);
+	sys_set_bit(config->base + UART_RQR, USART_RQR_RXFRQ_Pos);
+
+	/* Request next buffer */
+	evt.type = UART_RX_BUF_REQUEST;
+	if (data->async.cb) {
+		data->async.cb(dev, &evt, data->async.user_data);
+	}
+	uart_sf32lb_async_timer_start(&data->async.rx.timeout_work, timeout);
+
+unlock:
+	irq_unlock(key);
+	return err;
+}
+
+static int uart_async_sf32lb_rx_buf_rsp(const struct device *dev, uint8_t *buf, size_t len)
+{
+	struct uart_sf32lb_data *data = dev->data;
+
+	data->async.rx.next_buf = buf;
+	data->async.rx.next_len = len;
+
+	return 0;
+}
+
+static int uart_async_sf32lb_rx_disable(const struct device *dev)
+{
+	const struct uart_sf32lb_config *config = dev->config;
+	struct uart_sf32lb_data *data = dev->data;
+	unsigned int key = irq_lock();
```

Comment:
please avoid initializing vars using calls, it's also unclear locking is taken if buried here